### PR TITLE
[vcpkg] z_vcpkg_get_cmake_vars(): Use build-type-specific cache variable if VCPKG_BUILD_TYPE is defined

### PR DESCRIPTION
**Pull request by mkhon**
**Describe the pull request**

Some ports (e.g. gettext) execute debug and release builds manually (with different options). If a globally cached copy of cmake vars file is used, the second build type has no cmake vars defined (e.g. CFLAGS is empty) because generated and globally cached cmake vars file includes only cmake vars for the first build type.

For example,

1) gettext[tools] build first runs build_libintl_and_tools(BUILD_TYPE "release") first which causes cmake vars file to be generated with the following contents (the example is for x64-windows-static):

```
include("${CMAKE_CURRENT_LIST_DIR}/cmake-vars-x64-windows-static-rel.cmake.log")
```

2) then build_libintl_only(BUILD_TYPE "debug") is run, but cmake vars file is cached and debug build vars include() is not generated and used

- #### What does your PR fix?  
Currently, gettext[tools] windows static builds fail with the following error (because CFLAGS for debug are lost and empty):

```
-- Performing post-build validation
Invalid crt linkage. Expected Debug,Static, but the following libs had:

    C:\vcpkg\packages\gettext_x64-windows-static\debug\lib\intl.lib: Release,Static

To inspect the lib files, use:
    dumpbin.exe /directives mylibfile.lib
Found 1 post-build check problem(s). To submit these ports to curated catalogs, please first correct the portfile:
    C:\vcpkg\ports\gettext\portfile.cmake
-- Performing post-build validation done
```

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
all

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
N/A

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

**Comment by dg0yt** [comment]
Don't generalize what is done in gettext. It is special because building it takes really long, in particular the tools. Maybe it needs another tweek in gettext to get the cmake variables cached before touch `VCPKG_BUILD_TYPE`.

**Comment by mkhon** [comment]
> Don't generalize what is done in gettext. It is special because building it takes really long (on windows), in particular the tools. Maybe it needs another tweek in gettext to get the cmake variables cached before touching `VCPKG_BUILD_TYPE`.

this will require calling z_vcpkg_get_cmake_vars() from gettext portfile.cmake which I think is a greater evil than fixing caching in z_vcpkg_get_cmake_vars()

**Review by dg0yt — APPROVED** [review]
LGTM now.

**Review by LilyWangLL — APPROVED** [review]

File: scripts/cmake/z_vcpkg_get_cmake_vars.cmake [hunk 1]
```diff
@@ -39,8 +39,13 @@ function(z_vcpkg_get_cmake_vars out_file)
         message(FATAL_ERROR "${CMAKE_CURRENT_FUNCTION} was passed extra arguments: ${arg_UNPARSED_ARGUMENTS}")
     endif()
 
-    if(NOT DEFINED CACHE{Z_VCPKG_GET_CMAKE_VARS_FILE})
-        set(Z_VCPKG_GET_CMAKE_VARS_FILE "${CURRENT_BUILDTREES_DIR}/cmake-vars-${TARGET_TRIPLET}.cmake.log"
+    if(DEFINED VCPKG_BUILD_TYPE)
```
Comment (strega-nil-ms) [inline scripts/cmake/z_vcpkg_get_cmake_vars.cmake:42]:
```suggestion
    if(DEFINED VCPKG_BUILD_TYPE)
        set(cmake_vars_file "${CURRENT_BUILDTREES_DIR}/cmake-vars-${TARGET_TRIPLET}-${VCPKG_BUILD_TYPE}.cmake.log")
```

File: scripts/cmake/z_vcpkg_get_cmake_vars.cmake [hunk 2]
```diff
@@ -39,8 +39,13 @@ function(z_vcpkg_get_cmake_vars out_file)
         message(FATAL_ERROR "${CMAKE_CURRENT_FUNCTION} was passed extra arguments: ${arg_UNPARSED_ARGUMENTS}")
     endif()
 
-    if(NOT DEFINED CACHE{Z_VCPKG_GET_CMAKE_VARS_FILE})
-        set(Z_VCPKG_GET_CMAKE_VARS_FILE "${CURRENT_BUILDTREES_DIR}/cmake-vars-${TARGET_TRIPLET}.cmake.log"
+    if(DEFINED VCPKG_BUILD_TYPE)
+        set(cache_var "Z_VCPKG_GET_CMAKE_VARS_FILE_${VCPKG_BUILD_TYPE}")
+    else()
```
Comment (strega-nil-ms) [inline scripts/cmake/z_vcpkg_get_cmake_vars.cmake:44]:
```suggestion
    else()
        set(cmake_vars_file "${CURRENT_BUILDTREES_DIR}/cmake-vars-${TARGET_TRIPLET}.cmake.log")
```

File: scripts/cmake/z_vcpkg_get_cmake_vars.cmake [hunk 3]
```diff
@@ -39,8 +39,13 @@ function(z_vcpkg_get_cmake_vars out_file)
...
 
-    if(NOT DEFINED CACHE{Z_VCPKG_GET_CMAKE_VARS_FILE})
-        set(Z_VCPKG_GET_CMAKE_VARS_FILE "${CURRENT_BUILDTREES_DIR}/cmake-vars-${TARGET_TRIPLET}.cmake.log"
+    if(DEFINED VCPKG_BUILD_TYPE)
+        set(cache_var "Z_VCPKG_GET_CMAKE_VARS_FILE_${VCPKG_BUILD_TYPE}")
+    else()
+        set(cache_var Z_VCPKG_GET_CMAKE_VARS_FILE)
+    endif()
+    if(NOT DEFINED CACHE{${cache_var}})
+        set(${cache_var} "${CURRENT_BUILDTREES_DIR}/cmake-vars-${TARGET_TRIPLET}.cmake.log"
```
Comment (strega-nil-ms) [inline scripts/cmake/z_vcpkg_get_cmake_vars.cmake:48]:
```suggestion
        set(${cache_var}  "${cmake_vars_file}"
```

File: scripts/cmake/z_vcpkg_get_cmake_vars.cmake [hunk 4]
```diff
@@ -58,8 +63,8 @@ function(z_vcpkg_get_cmake_vars out_file)
         if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "debug")
             string(APPEND include_string "include(\"\${CMAKE_CURRENT_LIST_DIR}/cmake-vars-${TARGET_TRIPLET}-dbg.cmake.log\")\n")
         endif()
-        file(WRITE "${Z_VCPKG_GET_CMAKE_VARS_FILE}" "${include_string}")
+        file(WRITE "${${cache_var}}" "${include_string}")
```
Comment (strega-nil-ms) [inline scripts/cmake/z_vcpkg_get_cmake_vars.cmake:66]:
```suggestion
        file(WRITE "${cmake_vars_file}" "${include_string}")
```

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]
We want to make sure to not overwrite existing files.

**Comment by strega-nil-ms** [comment]
Thanks!

File: scripts/cmake/z_vcpkg_get_cmake_vars.cmake [hunk 5]
```diff
@@ -39,8 +39,15 @@ function(z_vcpkg_get_cmake_vars out_file)
         message(FATAL_ERROR "${CMAKE_CURRENT_FUNCTION} was passed extra arguments: ${arg_UNPARSED_ARGUMENTS}")
     endif()
 
-    if(NOT DEFINED CACHE{Z_VCPKG_GET_CMAKE_VARS_FILE})
-        set(Z_VCPKG_GET_CMAKE_VARS_FILE "${CURRENT_BUILDTREES_DIR}/cmake-vars-${TARGET_TRIPLET}.cmake.log"
+    if(DEFINED VCPKG_BUILD_TYPE)
+        set(cmake_vars_file "${CURRENT_BUILDTREES_DIR}/cmake-vars-${TARGET_TRIPLET}-${VCPKG_BUILD_TYPE}.cmake.log")
```
Comment (mkhon) [inline scripts/cmake/z_vcpkg_get_cmake_vars.cmake:43]: @strega-nil-ms I realized this change will lead to file name clash between x64-windows triplet "release" build and x64-windows-release triplet "all" build

Should better be changed to something like

set(cmake_vars_file "${CURRENT_BUILDTREES_DIR}/cmake-vars-${TARGET_TRIPLET}.${VCPKG_BUILD_TYPE}.cmake.log")

**Comment by mkhon** [comment]
@strega-nil-ms please see my last review comment. I realized there will still be a file name clash

**Comment by LilyWangLL** [comment]
@mkhon You need open a new PR to fix it.

**Comment by mkhon** [comment]
> @mkhon You need open a new PR to fix it.

See #22501

**Pull request merged**